packaging/metadata.py

```diff
@@ -0,0 +1,416 @@
+    }
+    _PARSING_POLICY: ClassVar[Policy] = EmailPolicy(max_line_length=math.inf, utf8=True)
+
+    @classmethod
+    def _canonical_field(cls, field: str) -> str:
+        words = _normalize_field_name_for_dynamic(field).split("-")
+        ucfirst = "-".join(w[0].upper() + w[1:] for w in words)
+        replacements = {"Url": "URL", "Email": "email", "Page": "page"}.items()
+        return reduce(lambda acc, x: acc.replace(x[0], x[1]), replacements, ucfirst)
+
+    @classmethod
+    def _parse_requires_python(cls, value: str) -> SpecifierSet:
+        if value and value[0].isnumeric():
+            value = f"=={value}"
+        return SpecifierSet(value)
+
+    @classmethod
+    def _parse_req(cls, value: str) -> Requirement:
+        try:
+            return Requirement(value)
+        except InvalidRequirement:
+            # Some old examples in PEPs use "()" around versions without an operator
+            # e.g.: `Provides: xmltools (1.3)`
+            name, _, rest = value.strip().partition("(")
+            value = f"{name}(=={rest}"
+            return Requirement(value)
+
+    @classmethod
+    def _convert_single_req(cls, value: Union[Requirement, str]) -> Requirement:
+        return value if isinstance(value, Requirement) else cls._parse_req(value)
+
+    @classmethod
+    def _convert_emails(
+        cls, value: Collection[Union[str, Tuple[str, str]]]
+    ) -> Iterator[EmailAddress]:
+        for email in value:
+            if isinstance(email, str):
+                yield from cls._parse_emails(email)
+            elif isinstance(email, tuple) and email[1]:
+                yield EmailAddress(email[0], email[1])
+
+    @classmethod
+    def _parse_emails(cls, value: str) -> Iterator[EmailAddress]:
+        singleline = " ".join(value.splitlines())
+        if singleline.strip() == "UNKNOWN":
+            return
+        address_list = AddressHeader.value_parser(singleline)
+        for mailbox in address_list.all_mailboxes:
+            yield EmailAddress(mailbox.display_name, mailbox.addr_spec)
+
+    @classmethod
+    def _unescape_description(cls, content: str) -> str:
+        """Reverse RFC-822 escaping by removing leading whitespaces from content."""
+        lines = content.splitlines()
+        if not lines:
+            return ""
+
+        first_line = lines[0].lstrip()
+        text = textwrap.dedent("\n".join(lines[1:]))
+        other_lines = (line.lstrip("|") for line in text.splitlines())
+        return "\n".join(chain([first_line], other_lines))
+
+    def _validate_dynamic(self, normalized: str) -> bool:
+        field = normalized.lower().replace("-", "_")
+        if not hasattr(self, field):
+            raise InvalidCoreMetadataField(normalized)
+        if field in self._NOT_DYNAMIC:
+            raise InvalidDynamicField(normalized)
+        if getattr(self, field):
+            raise StaticFieldCannotBeDynamic(normalized)
+        return True
+
+
+class ImmutableMapping(Mapping[A, B]):
```

Comment:
I don't think it's worth going so far as to create our own immutable mapping class to make the entire set of data immutable. The idea of using an immutable dataclass was to prevent accidental race conditions / inconsistencies around `dynamic`, not to specifically make sure the data itself was immutable.

Reply:
I confess that I went a bit overboard with the whole frozen/immutable thing. Thank you very much for spotting that out.